### PR TITLE
feat(ingest/db2): add source for IBM Db2

### PR DESCRIPTION
**Pull request by ligfx**
Supports:
- Tables
- Views
- Stored procedures
- Lineage from views and stored procedures (including Db2's special rules around looking up unqualified names)
- Comments on tables, views, columns, and stored procedures

One thing to note: spinning up Db2 for LUW for tests is slooooow 😞 A full test cycle takes almost 3 minutes on my computer.

Tested against:
- Db2 for LUW (freely available Docker containers, can run on AWS, etc.). Integration tests included in PR.
- Db2 for IBM i (AS/400) using [PUB400.COM](https://pub400.com/) and the separate [IBM i Access ODBC Driver](https://ibmi-oss-docs.readthedocs.io/en/latest/odbc/installation.html). Manual testing only.

Has nominal support for Db2 for z/OS but will rely on community to see if that actually works.

Screenshot 2025-10-09 at 12 38 08 PM

Remaining chores:
- [x] Update ibm_db_sa dependency to 0.4.3
- [x] Validate `database` is not none (is this required for AS400?)
- [x] Unit test z/OS PATHSCHEMAS regex
- [x] Cleanup/remove statement splitting logic
- [x] Assign integration test batch
- [x] Add comment about custom SQLGlot dialect
- [x] Add notes on manual testing AS400
- [x] Clean up get_view_default_db_schema comment, ask about logic

### CODE REVIEW

**Comment by sgomezvillamor** [comment]
> Db2 for IBM i (AS/400) using [PUB400.COM](https://pub400.com/) and the separate [IBM i Access ODBC Driver](https://ibmi-oss-docs.readthedocs.io/en/latest/odbc/installation.html). Manual testing only.

Do we have internal instructions on how to do this manual testing.

Also, it would be nice to have some internal doc in https://www.notion.so/acryldata/a00c812926f544d7bfe46d4d8154ba07?v=843815ea3b724a2b8dba0dac2b794d70 At least, track the new source :-)

**Comment by sgomezvillamor** [comment]
@cursoragent can you review this PR?

File: metadata-ingestion/docs/sources/db2/db2_post.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+This requires you to [install the IBM i Access ODBC Driver](https://ibmi-oss-docs.readthedocs.io/en/latest/odbc/installation.html) along with an ODBC driver manager such as `unixodbc`.
+
+To use the IBM i Access ODBC Driver rather than the default driver, specify the `db2+pyodbc400` scheme in your ingestion configuration:
+
+```yaml
+scheme: "db2+pyodbc400"
+```
+
+If you're using [DataHub Cloud's Remote Executor](https://docs.datahub.com/docs/managed-datahub/remote-executor/about), you can either build a custom image with the ODBC driver included, or sideload the driver. For instance, to sideload the driver and when using the Remote Executor Helm chart, you might use the following Helm values file:
```
Comment (sgomezvillamor) [inline metadata-ingestion/docs/sources/db2/db2_post.md:13]: Remote executor is datahub cloud only, so we shouldn't mention this here

IMO this docs should become actual udpates in our remote executor docker image, AFAIK we already install some odbc drivers there and we should add the new ones

In that case, we may just document here the backend dependency requirement for this connector to work with proper driver

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    # to get further metadata on them).
+    # https://github.com/ibmdb/python-ibmdbsa/issues/153
+    # https://github.com/ibmdb/python-ibmdbsa/issues/170
+    dialect._reflector.normalize_name = lambda name: name
+    dialect._reflector.denormalize_name = lambda name: name
+
+
+class Db2Config(BasicSQLAlchemyConfig):
+    database: Optional[str] = pydantic.Field(
+        default=None, description="The Db2 database to ingest from."
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/sql/db2.py]: what if None the default? everything is ingested?

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+                argument_signature=None,
+                return_type=None,
+                extra_properties={},
+            )
+
+
+def _db2_get_view_qualifier_quoted(
+    inspector: Inspector, schema: str, view: str
+) -> Optional[str]:
+    if inspector.has_table("VIEWS", schema="SYSCAT"):
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/sql/db2.py:180]: we may define an enum for: Db2 LUW, Db2 z/OS and Db2 i/AS400

and a method: def _identify_db2_type(inspector): enum

because I would expect that to become a recurrent pattern

File: metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py [hunk 4]
```diff
@@ -1155,6 +1155,13 @@ def _get_view_definition(self, inspector: Inspector, schema: str, view: str) ->
         except NotImplementedError:
             return ""
 
+    def get_view_default_db_schema(
+        self, dataset_name: str, _inspector: Inspector, _schema: str, _view: str
+    ) -> Tuple[Optional[str], Optional[str]]:
+        # Some databases use different implicit databases/schemas for unqualified
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/sql/sql_common.py]: "some databases"?
this method consistently returns None, None, how that applies to some databases and which are those databases?

File: metadata-ingestion/src/datahub/sql_parsing/sql_parsing_common.py [hunk 5]
```diff
@@ -7,6 +7,7 @@
 
 PLATFORMS_WITH_CASE_SENSITIVE_TABLES = {
     "bigquery",
+    "db2",
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/sql_parsing/sql_parsing_common.py:10]: btw, do we have the logo for this new platform?

File: metadata-ingestion/tests/integration/db2/db2_basic_mces_golden.json [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+    }
+},
+{
+    "entityType": "container",
+    "entityUrn": "urn:li:container:c5f650538201e34d961c8d9e24ac94e4",
+    "changeType": "UPSERT",
+    "aspectName": "dataPlatformInstance",
+    "aspect": {
+        "json": {
+            "platform": "urn:li:dataPlatform:db2"
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/integration/db2/db2_basic_mces_golden.json:152]: can we test also with platform instance?

File: metadata-ingestion/tests/integration/db2/test_db2.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    return pytestconfig.rootpath / "tests/integration/db2"
+
+
+def is_db2_up(container_name: str) -> bool:
+    cmd = f"docker logs {container_name} 2>&1 | grep 'Setup has completed.'"
+    ret = subprocess.run(
+        cmd,
+        shell=True,
+    )
+    return ret.returncode == 0
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/integration/db2/test_db2.py:31]: I usually prefer to push down this to the docker infra level.
You can define a healthcheck in the docker compose to determine the service is healthy and so you don't need to handle this in the app layer

File: metadata-ingestion/tests/integration/db2/test_db2.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        wait_for_port(
+            docker_services,
+            "testdb2",
+            DB2_PORT,
+            timeout=600,
+            checker=lambda: is_db2_up("testdb2"),
+        )
+
+        setup_filename = test_resources_dir / "setup" / "setup.sql"
+        statements = _split_statements(open(setup_filename).read())
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/integration/db2/test_db2.py:73]: have you check if you can initialize the execution of the setup.sql at docker level?

alternatively, why splitting statements instead of having statements already pre-split (eg, one per line, or json array)?

File: metadata-ingestion/tests/integration/db2/test_db2.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    "config_filename",
+    [
+        "db2_basic.yml",
+        "db2_case_sensitivity.yml",
+        "db2_comments.yml",
+        "db2_procedures.yml",
+        "db2_view_qualifier.yml",
+    ],
+)
+@pytest.mark.integration
```
Comment (sgomezvillamor) [inline metadata-ingestion/tests/integration/db2/test_db2.py:96]: we may need to assign some batch for the integration tests

File: metadata-ingestion/setup.py [hunk 10]
```diff
@@ -462,6 +462,8 @@
     "datahub-lineage-file": set(),
     "datahub-business-glossary": set(),
     "delta-lake": {*data_lake_profiling, *delta_lake},
+    # TODO: remove ibm_db_sa git URL once v0.4.3 is released
```
Comment (sgomezvillamor) [inline metadata-ingestion/setup.py]: so you did some contribs that hasn't been published yet?

**Review by sgomezvillamor — APPROVED** [review]
Overall LGTM and no red flag, so approving

I left some comments and also cursor left some interesting suggestions that may be considered

**Comment by ligfx** [comment]
> Do we have internal instructions on how to do this manual testing.

No, where would be a good place to put that?

> Also, it would be nice to have some internal doc in https://www.notion.so/acryldata/a00c812926f544d7bfe46d4d8154ba07?v=843815ea3b724a2b8dba0dac2b794d70 At least, track the new source :-)

Added!

File: metadata-ingestion/docs/sources/db2/db2_post.md (hunk 1 above)
Comment (ligfx) [reply metadata-ingestion/docs/sources/db2/db2_post.md:13]: > Remote executor is datahub cloud only, so we shouldn't mention this here

Disagree, if specific configuration is needed for Remote Executor we need to document that. We can remove this documentation after the ODBC driver is added into the Remote Executor image.

File: metadata-ingestion/src/datahub/sql_parsing/sql_parsing_common.py (hunk 5 above)
Comment (ligfx) [reply metadata-ingestion/src/datahub/sql_parsing/sql_parsing_common.py:10]: Yep, it was added last year: https://github.com/datahub-project/datahub/pull/10601

File: metadata-ingestion/tests/integration/db2/test_db2.py (hunk 7 above)
Comment (ligfx) [reply metadata-ingestion/tests/integration/db2/test_db2.py:31]: I tried a couple ways to do a healthcheck inside the Docker service, but couldn't figure it out 😞  The CLI tools are rather obtuse.

File: metadata-ingestion/tests/integration/db2/test_db2.py (hunk 9 above)
Comment (ligfx) [reply metadata-ingestion/tests/integration/db2/test_db2.py:96]: What do you mean by "assign some batch"?

File: metadata-ingestion/setup.py (hunk 10 above)
Comment (ligfx) [reply metadata-ingestion/setup.py]: Yep, v0.4.3 will have a number of bug fixes.

File: metadata-ingestion/tests/integration/db2/test_db2.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+        if "BEGIN" in statement.upper():
+            inside_begin = True
+        if statement.upper() == "END":
+            inside_begin = False
+    return statements
+
+
+@pytest.fixture(scope="module")
+def db2_runner(docker_compose_runner, pytestconfig, test_resources_dir):
+    with docker_compose_runner(
```
Comment (ligfx) [inline metadata-ingestion/tests/integration/db2/test_db2.py:61]: @sgomezvillamor it would be nice to be able to skip Docker Compose and repoint this at an existing database service, for faster local testing and i/AS400 testing. Do we have any existing patterns for that? Maybe an environment variable?

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py (hunk 2 above)
Comment (ligfx) [reply metadata-ingestion/src/datahub/ingestion/source/sql/db2.py]: No, `database` must be specified. Is there a way to specify that here? I tried making it non-optional, but Pydantic complained since the base config model has it as optional.

Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/sql/db2.py]: if you cannot set it non-optional, then you may add a pydantic `field_validator` that may fail if `None`

File: metadata-ingestion/tests/integration/db2/test_db2.py (hunk 7 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/integration/db2/test_db2.py:31]: yes, that sometimes happen
👌

File: metadata-ingestion/tests/integration/db2/test_db2.py (hunk 9 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/integration/db2/test_db2.py:96]: https://github.com/search?q=repo%3Adatahub-project%2Fdatahub%20%22pytest.mark.integration_batch%22&type=code

File: metadata-ingestion/tests/integration/db2/test_db2.py (hunk 11 above)
Comment (sgomezvillamor) [reply metadata-ingestion/tests/integration/db2/test_db2.py:61]: AFAIK we have never done something similar

**Comment by sgomezvillamor** [comment]
> No, where would be a good place to put that?
> 

Maybe some README.md in the integration tests directory for db2?

File: metadata-ingestion/tests/integration/db2/test_db2.py (hunk 8 above)
Comment (ligfx) [reply metadata-ingestion/tests/integration/db2/test_db2.py:73]: Looked into it, the Docker image doesn't support running SQL upon startup, unfortunately. Having the statements as text is much easier for maintenance than pre-splitting them.

File: metadata-ingestion/tests/integration/db2/db2_basic_mces_golden.json (hunk 6 above)
Comment (ligfx) [reply metadata-ingestion/tests/integration/db2/db2_basic_mces_golden.json:152]: Is there something specific about this source you think would make platform instances not work? All of the platform instance logic is in the parent class, `SQLAlchemySource`.

File: metadata-ingestion/docs/sources/db2/db2_post.md [hunk 12]
```diff
@@ -0,0 +1,52 @@
+### Using the IBM i Access ODBC Driver
+
+This source can connect to Db2 for IBM i (AS/400) directly using the IBM i Access ODBC Driver rather than using the default CLI driver that requires Db2 Connect.
+
+This requires you to [install the IBM i Access ODBC Driver](https://ibmi-oss-docs.readthedocs.io/en/latest/odbc/installation.html) along with an ODBC driver manager such as `unixodbc`.
```
Comment (sgomezvillamor) [inline metadata-ingestion/docs/sources/db2/db2_post.md:5]: to confirm my understanding: we are not installing this in our CI github workflows, so we are not testing this IBM i Access ODBC Driver. Am I right?

File: metadata-ingestion/docs/sources/db2/db2_pre.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+uri_args:
+  Security: SSL
+```
+
+### Db2 Platform Support
+
+This source has been tested against:
+
+- Db2 for LUW using the default CLI driver (included)
+- Db2 for IBM i (AS/400) using the IBM i Access ODBC Driver (not included, see [Using the IBM i Access ODBC Driver](#using-the-ibm-i-access-odbc-driver) below)
```
Comment (sgomezvillamor) [inline metadata-ingestion/docs/sources/db2/db2_pre.md]: what does "included"/"not included" mean here?

File: metadata-ingestion/docs/sources/db2/db2_pre.md [hunk 14]
```diff
@@ -0,0 +1,32 @@
+### Permissions
+
+In order to execute this source the user needs `SELECT` privileges on the system schema and tables:
+
+- Db2 for LUW: `SYSCAT.*`
+- Db2 for z/OS: `SYSIBM.*`
+- Db2 for IBM i (AS/400): `QSYS2.*`
+
+Additionally, when profiling is enabled, the user will need `SELECT` privileges on the tables to be profiled.
```
Comment (sgomezvillamor) [inline metadata-ingestion/docs/sources/db2/db2_pre.md:9]: Does that lineage require some additional permission?

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+        "Specify regex to match the entire procedure name in schema.procedure_name format.",
+    )
+
+    uri_args: Dict[str, str] = pydantic.Field(
+        default={},
+        description="Arguments to add to the URL when connecting.",
+    )
+
+    # Override defaults
+    host_port: str = pydantic.Field(default="localhost:50000")
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/sql/db2.py]: even if obvious, we could add some description

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+    scheme: str = pydantic.Field(
+        default="db2",
+        description="ibm_db_sa scheme to use (db2, db2+pyodbc, db2+pyodbc400).",
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/sql/db2.py:88]: I'm very little familiar with DB2 and these schema differences

For non db2 expert users that want to integrate their system with DataHub or even for those non db2 experts that will maintain this source 😅 , we could add some explanation about these schemas:

This is what I found in a quick search

> the choice of ibm_db_sa scheme depends on the specific IBM Db2 database you are connecting to and the desired connection method.
> - `db2`: This scheme uses the ibm_db driver directly, which is the native Python driver for IBM Db2 databases. This is the recommended choice for connecting to Db2 for LUW (Linux, Unix, Windows), Db2 on Cloud, and Db2 for z/OS when using the ibm_db driver.
> - `db2+pyodbc`: This scheme utilizes pyodbc as an intermediary to connect to Db2 via an ODBC driver. This is suitable when you need to use an existing ODBC driver for your Db2 instance, or if you are connecting to a Db2 version or platform where the ibm_db driver might not be the primary choice, and an ODBC driver is readily available and configured.
> - `db2+pyodbc400`: This scheme is specifically designed for connecting to IBM Db2 for i (formerly AS/400) using pyodbc and the IBM i Access ODBC Driver. The 400 suffix explicitly indicates the use of the iSeries ODBC driver, ensuring proper handling of the platform's specific characteristics.

Not sure how much accuratted is this and likely the config is not the best place to have this detailed doc; it would be nice to have it somewhere.

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+@capability(SourceCapability.DOMAINS, "Supported via the `domain` config field")
+@capability(SourceCapability.DATA_PROFILING, "Optionally enabled via configuration")
+@capability(
+    SourceCapability.DELETION_DETECTION, "Enabled by default via stateful ingestion"
+)
+class Db2Source(SQLAlchemySource):
+    def __init__(self, config: Db2Config, ctx: PipelineContext):
+        super().__init__(config, ctx, "db2")
+        # register custom SQLGlot dialect
+        if not sqlglot.Dialect.get("db2"):
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/sql/db2.py:121]: having or not the db2 dialect, does that depend on the sqlglot version, or how is that?

File: metadata-ingestion/docs/sources/db2/db2_post.md (hunk 12 above)
Comment (ligfx) [reply metadata-ingestion/docs/sources/db2/db2_post.md:5]: Correct. We also don't have an i (AS/400) server to test against, unless we want to set up something with PUB400.COM internally that runs less often.

File: metadata-ingestion/docs/sources/db2/db2_pre.md (hunk 13 above)
Comment (ligfx) [reply metadata-ingestion/docs/sources/db2/db2_pre.md]: "Included" means that the default CLI driver is included with the `acryl-datahub[db2]` package via the `ibm_db` package, so there's no additional work needed.

"Not included" means that the IBM i Access ODBC Driver must be installed separately by the end user.

File: metadata-ingestion/docs/sources/db2/db2_pre.md (hunk 14 above)
Comment (ligfx) [reply metadata-ingestion/docs/sources/db2/db2_pre.md:9]: Which lineage?

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py (hunk 3 above)
Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/sql/db2.py:180]: I still think having a method for doing the discrimination is a good pattern

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py (hunk 16 above)
Comment (ligfx) [reply metadata-ingestion/src/datahub/ingestion/source/sql/db2.py:88]: The `db2+pyodbc400` scheme is documented in the docs changes for what you need to do to connect to a Db2 for i (AS/400) instance.

I'm not sure when someone would use the `db2+pyodbc` driver, or even if it works, but it's provided by the ibm_db_sa driver. I consider this as an expert option.

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+            (schema,),
+        )
+
+    else:
+        raise NotImplementedError(
+            "Couldn't find SYSCAT.ROUTINES, SYSIBM.SYSROUTINES, or QSYS2.SYSROUTINES"
+        )
+
+
+def _split_zos_pathschemas(text):
```
Comment (sgomezvillamor) [inline metadata-ingestion/src/datahub/ingestion/source/sql/db2.py]:
```suggestion
def _split_zos_pathschemas(text) -> List[str]:
```
?

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py (hunk 17 above)
Comment (ligfx) [reply metadata-ingestion/src/datahub/ingestion/source/sql/db2.py:121]: There is no Db2 SQLGlot dialect currently. I put this in here to be forward-compatible if SQLGlot ever does add a dialect. I'll add a comment.

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py (hunk 3 above)
Comment (ligfx) [reply metadata-ingestion/src/datahub/ingestion/source/sql/db2.py:180]: I don't agree on this one. Adding another method adds an additional layer of indirection and something to break, and we'd have to decide upon a way to do the discrimination (check the underlying SQLAlchemy inspector reflector type?). The current way is straightforward and doesn't have any business logic issues: if table X exists, use it.

**Review by sgomezvillamor — COMMENTED** [review]
Overall LGTM 
Most of my comments are about docs

Test coverage may be a little low, unfortunately I cannot see in codecov uncovered sources

Screenshot 2025-12-05 at 11 02 52

File: metadata-ingestion/src/datahub/ingestion/source/sql/db2.py (hunk 3 above)
Comment (sgomezvillamor) [reply metadata-ingestion/src/datahub/ingestion/source/sql/db2.py:180]: I'm happy to have comments resolved, but I'd appreciate a quick reply when resolving.

My concern is mainly about preventing this pattern from spreading throughout the codebase. In my opinion, the table-checking logic does encode business knowledge about the three Db2 variants (LUW, z/OS, i/AS400), and if we need to identify the Db2 type in multiple places, we'd be duplicating that logic. My view may be biased by the comments on every case of the if, which emphasize the "business" knowledge.

This is clearly opinionated territory and so not blocking. Resolving and leaving this to your final judgement. Thanks for engaging on this.

Comment (ligfx) [reply metadata-ingestion/src/datahub/ingestion/source/sql/db2.py:180]: Sorry about that, you had approved last time and this sounded like an optional suggestion so I didn't think it needed a response. I'll make sure they all get a quick reply in the future.

Fwiw, here's what Claude says:

> I'm writing some code to integrate with an external database. This database may have a few different variants, each with a different table I'd need to connect to. I have two options for writing this logic:
> 1. Run some tests to decide which variant is connected, and then access that table.
> 2. Check which table exists, then access that table.
>
> Any thoughts? What would you call this problem or these options?

> This is a classic pattern in database programming. I'd call it **schema detection** or **schema discovery**, and the two approaches you're describing are:
> 1. **Environment detection** - Figure out what you're dealing with first (test the environment, check version numbers, look at config), then act accordingly
> 2. **Direct probing** or **existence checking** - Just try to find what you need and work with what's there
>
> Most experienced developers would lean toward option 2 (checking if the table exists) for a few reasons:
> **It's more robust.** You're checking exactly what you need rather than inferring it from other signals. What if the variant detection logic has edge cases you didn't anticipate? What if someone manually modified the database?
> **It's simpler.** One query to check if a table exists is usually cleaner than running multiple tests to determine the variant, especially if those variant rules might change.
> **It's more maintainable.** If a new variant comes along, you just add another table to check. You don't need to update variant detection logic.
> The main exception would be if checking for table existence is expensive or if you need to know the variant for other reasons beyond just picking a table. Or if the variant detection is trivial (like, it's literally in a config file you already have).
> In practice, you might even combine them - check a config or environment variable first if available, but fall back to existence checking if needed. That gives you the best of both worlds.

**Comment by ligfx** [comment]
Thanks for the review @sgomezvillamor . I'd also especially love your thoughts on the `get_view_default_db_schema` method in the SQLAlchemySource class.

Context:
- Most SQL sources assume the default catalog/schema for resolving references in views is the same catalog/schema that the view itself lives in.
- Db2 handles this differently, and stores an explicit schema path that references should be looked up in.

Code:
- SQLAlchemySource grabs the catalog/schema by parsing it back out from the dataset name string, which doesn't feel great since we already had that information at one point.
- The Db2 source needs to go grab the information from a system table, so needs the raw schema and table identifiers, exactly as returned by Db2.
- I originally added `get_view_default_db_schema` as an _alternative_ to parsing the catalog/schema from the dataset name, but that doesn't feel good either.
- Might be worth making `get_view_default_db_schema` wrap the `get_db_schema(name)` logic? Though that method would require the raw catalog, schema, and name, as well as the generated name, which gets verbose. I could also plumb the raw names through and remove the current `get_db_schema(name)` logic from view parsing entirely (probably the cleanest engineering option).

**Comment by sgomezvillamor** [comment]
> Though that method would require the raw catalog, schema, and name, as well as the generated name, which gets verbose. 

That means adding those parameters to the `get_db_schema`? parameters that would be ignored by all sources but db2 one, right? 

That would make sense if we have expectation that some other source will have similar constraint.

> I could also plumb the raw names through and remove the current `get_db_schema(name)` logic from view parsing entirely (probably the cleanest engineering option).

That would have no impact on any other source different than db2. So looks clean.

File: metadata-ingestion/docs/sources/db2/db2_pre.md (hunk 14 above)
Comment (sgomezvillamor) [reply metadata-ingestion/docs/sources/db2/db2_pre.md:9]: SELECT privilege on the system metadata tables (SYSCAT, SYSIBM, or QSYS2) is required for metadata extraction. I’m wondering if this also covers lineage, or if we need access to additional system tables. Just double-checking, since other sources sometimes require extra tables or schemas.

File: metadata-ingestion/docs/sources/db2/db2_post.md (hunk 12 above)
Comment (sgomezvillamor) [reply metadata-ingestion/docs/sources/db2/db2_post.md:5]: Got it, thanks!

File: metadata-ingestion/docs/sources/db2/db2_pre.md (hunk 13 above)
Comment (sgomezvillamor) [reply metadata-ingestion/docs/sources/db2/db2_pre.md]: Got it. Thanks

Something like this would make everything clear

```
- Db2 for LUW using the default CLI driver (already included when installing the `acryl-datahub[db2]` package via the `ibm_db` package)
- Db2 for IBM i (AS/400) using the IBM i Access ODBC Driver (not included and so must be installed separately by the end user; see [Using the IBM i Access ODBC Driver](#using-the-ibm-i-access-odbc-driver) below)
```

File: metadata-ingestion/docs/sources/db2/db2_pre.md (hunk 14 above)
Comment (ligfx) [reply metadata-ingestion/docs/sources/db2/db2_pre.md:9]: No additional privileges needed, it's all in those system metadata tables.

**Comment by ligfx** [comment]
> > Though that method would require the raw catalog, schema, and name, as well as the generated name, which gets verbose.
> 
> That means adding those parameters to the `get_db_schema`? parameters that would be ignored by all sources but db2 one, right?
> 
> That would make sense if we have expectation that some other source will have similar constraint.
> 
> > I could also plumb the raw names through and remove the current `get_db_schema(name)` logic from view parsing entirely (probably the cleanest engineering option).
> 
> That would have no impact on any other source different than db2. So looks clean.

Went with just passing the generated name through to `get_view_default_db_schema` and parsing the catalog and schema back out.

I wanted to plumb the raw catalog and schema names through and do it cleanly, but the MySQL and other `TwoTierSQLAlchemySource` classes are doing weird things with database vs. schema naming and I wasn't confident in not breaking things.

(IMO the SQL common logic should be cleaned up to talk about "catalogs" and "schemas" instead. "database" is not a standard SQL term and can mean different things depending on the source system, hence the complexity in the MySQL source/other sources where they're trying to fit a model where "database" and "schema" are synonyms into the common model where "database" and "schema" are different things.)

Ready for final review.

**Review by sgomezvillamor — APPROVED** [review]
LGTM
